### ISSUE TITLE
Should infinite Grids be allowed

### ISSUE THREAD

**Issue by lovesegfault** [bemeurer/hemoglobin#31]
On #30 the structure of Grid was defined as such
```rust
pub struct Grid {
    grid: CellSet,
    bounds: Option<(usize, usize)>,
}
```
Here, the choice for `Option<(usize, usize)>` for the bounds was due to the fact that in conversation with @strilanc last week he pointed out that since the grid was a `HashSet`, it could effectively be treated as infinite if we so wished. In this case we would only view a portion of the grid, but all of it would be generated and iterated upon.

Do we want to support this? Or should this possibility be deprecated and `Grid` be required to have boundaries?

**Comment by maffoo** [comment]
Setting the bounds to `(usize::MAX, usize::MAX)` gives an effectively infinite `Grid` (as big as is allowed by the `Cell` coordinates themselves (if you interpret the bounds as the max allowed row/column, rather than the number of rows/columns). This to me feels more natural than letting the bounds be optional.

**Comment by lovesegfault** [comment]
Well, the coordinates could very well be `BigUint` to allow for a virtually infinite `Grid`, they're `usize` right now just because that type was convenient I suppose. I guess the real question is should we bother? Should we care about supporting a `Grid` *larger* than `usize::MAX` (~1.84*10^19)?

I think maybe it could be worth it to try using `BigUint` coordinates, generating some obnoxiously large worlds, and benchmarking to see what happens. Perhaps (and specially once we parallelize the load) it may be feasible, although my guess is that it will just take forever to compute anything.

**Comment by mion** [comment]
I see at least two advantages of restructuring the grid to not include bounds at all:

1. You can simplify the update algorithms (you don't have to check for borders).
2. You avoid allocating all the unnecessary space. This may not be much of a big deal since I have no idea how smart the Rust compiler is at this point; perhaps it's clever enough to optimize it somehow?

Using a very large integer to simulate an infinite grid feels conceptually wrong for me. If you refactor the code to abstract away the access to the cells, you can remove the boundaries from the data structure and the update/rendering algorithms would remain the same.

**Comment by DanielSank** [comment]
@mion If there are no bounds then can't the automaton can use up arbitrarily large amounts of memory? If so, that sounds like a bug to me. I think there needs to be a (optional) way to limit memory use.

**Comment by lovesegfault** [comment]
@mion I don't think it's interesting *force* the grid to be boundless. I am, however, all for having it be boundless as an option. With regards to your points, I think boundary checks are a nice idea assuming the grid has boundaries, and since the grid is represented as a HashSet, no allocation of unused space is made. 

The grid is just a set of points with living cells, and an `Option<(usize, usize)>` to represent the bounds, where `None` means no bounds.  When  the `Grid` is instantiated with bounds passed into it we [preallocate the HashSet to an appropriate size](https://github.com/bemeurer/hemoglobin/blob/master/src/lib.rs#L25), and when that's not possible we leave it to the language to allocate it's default value and then reallocate later on if necessary.

Finally, if by "very large integer" you mean a `BigUint`, we only use that as a representation for the [Rule](https://github.com/bemeurer/hemoglobin/blob/master/src/lib.rs#L111).

**Comment by lovesegfault** [comment]
@DanielSank Well, sure, on an infinite grid, given enough time, the automaton would use all available memory, because the `HashSet` would grow so large. At that point we would start swapping until we also use up all swap space and then we would crash. I think this is a feature, not a bug though.

It just means "The only limiting factor to how large a `Grid` you can have is how much memory you have", which IMO is desirable.

**Comment by mion** [comment]
@DanielSank Indeed! But if such is the nature of some automata then wouldn't it be more interesting if, for the sake of practice (which I think is the goal of this project, right @bemeurer?), we let it grow freely and then study some way to address it from the realm of Rust? As a language it seems to excel at this low level stuff, so it's probably a nice learning opportunity.

@bemeurer From an architectural point of view, the poor automaton doesn't need to know it's living inside a bounded rectangle — the memory issue is a computer problem and these beloved mathematical beings just care about reproducing. They are defined by currently living cells, neighbors of living cells and rules; they are boundless by definition. Though there is probably some class of crazy mathematicians studying what happens when one adds rules that take boundaries as input. (Which is rather interesting, come to think of it... is our universe like this?)

From a performance point of view, it is also running in `O( width * height )` needlessly, right? I'm looking at the step function and assuming every rule runs in `O(1)`:
https://github.com/bemeurer/hemoglobin/blob/b1e4d6554c7bc5aa46d59e090e946d280159744f/src/lib.rs#L160-L167

I've written a cellular automata simulator a while ago and if I remember correctly, the pseudocode for my `step` function was something like this:
```
updates = []
for( c in living_cells )
  neighs = neighbors(c)
  rule = rules.match(neighs)
  update = rule.apply(c, neighs)
  updates.add(update)
for( update in updates )
  living_cells.apply(update)
```

I feel this abstraction fits the definition very well. I'm trying to think of a drawback of not tying the code to a two-dimensional matrix but I'm not finding any. I did come up with another interesting benefit for this abstraction though, which is you can easily implement cellular automata that are based on hexagonal cells or even continuous automata.
